### ISSUE TITLE
Servers using a port which is not default can't announce to central server ?

### ISSUE THREAD

**Issue by hartym** [corrados/jamulus#91]
Hello,

Please forgive me if I misunderstood how things work here, I'm trying to setup more than one jamulus server on one physical box, using different ports. It looks like only the server running on default port can announce to the central server. I tried to read the code but I fear my cpp skills won't do it without investing a lot of time here, looks like "-p port" option works but not for announcement which would use LLCON_DEFAULT_PORT_NUMBER ?

I can try to help if I have the right pointers.

Thanks

**Comment by pljones** [comment]
When you connect domestically to the internet, you usually do so by a router with a firewall which also manages a feature called network address translation.

This is mainly used to ensure that a request from a computer on your local area network has its IP address translated to the global IP address that identifies the router, so that a response is sent back to the right place.

In addition to the IP address, though, the port number can need to be changed - two PCs might be trying to connect outward using the same number, so the router has to do magic with that as well.

Internally, though, the computers are not aware of any of this.  You can set up a server that says "I am 192.168.3.47 and I'm listening on port 54321 for Jamulus data".  Globally, though, that gets you nowhere -- another computer on your network might have already had some random port it was using translated to 54321 and thus the router can't just go ahead and use it.

Further, no one on the global network can say "Hey, let's call up the server on 192.168.3.47" -- because that address is explicitly restricted to local use only - it won't ever travel over the global networks.

All of that needs to be understood.

So what do you do?  You router, remember, is already doing masses of translations.  So it needs to be told a bit more information.  Let's say your global IP address (Google for "what's my IP") is 8.8.4.4 (it isn't, Google own that ;) ).  That gets people as far as your router already.  But your router needs also to be told "If someone from the outside world connects on <some port here> send them to <this internal computer IP> and <this port number on that IP>".

`-p` just overrides the default.  It doesn't change behaviour of the above at all.

The "announcement" is made simply by specifying the address of the "Central Server".  When you connect, your router address and port are then published on the list.  You control the port by setting up the router (not directly with -p, unless you've got a router set up some fancy way I've never met).

Now, also note that the global address on your router can change.  Domestically, this can vary between frequently (I call daily frequently) and rarely.  You can, possibly, request "fixed IP" for your line, which can make life easier -- but the Central Server gets its details updated _very_ frequently (a few times an hour), so it's often not a problem.

**Comment by hartym** [comment]
Hello, thanks for your extensive answer.

I'm not using a domestic connection, I run the servers on one dedicated box in a datacenter. Currently, there are two : 51.158.106.80:22124 and 51.158.106.80:1234. They both can be connected on the public internet network, so I can tell there are no routing issue.

While the first one is correctly annnounced to the central server, the second one never show up. 

Both servers are run using the same arguments, except for the -p which is set to 1234 for the second one. I guessed that the "announcement" process ignored the port which would explain this behavior but maybe the reason differs.

Thanks

**Comment by pljones** [comment]
First, make sure you have the very latest version.  It's got a useful update.

Then, if you can, run the server with the UI (on a local X session, if needs be).  Use that to replicate the connection (you only need the "jamulus -s" - skip the "-n").

That should tell whether the servers are _trying_ to connect and what the outcome is.

**Comment by hartym** [comment]
Ok so, I did compile 3.5.1 to switch to latest (from git tag). Then setup a local X server and ssh -X to the box to run a server with GUI.

I managed to register it from another port (but I have good reason to think 3.5.0 would have done the same).

At some point though, it told me "Central Server Full", which is a very good explanation for a server not showing up.

Do you think it would be possible to add this information as a log in the console ? Would be really helping for headless servers. If you point me to the related files, I can try to propose a merge request.

Also, is there any plans to increase the central server(s) capabilities on this regard ? I guess lockdown brought a lot of activity there :)

Thanks again and thanks for this great software.

**Comment by pljones** [comment]
There's a second central server - in the UI there's a drop down and on the command line, there's an option.

I guess there could be an additional "--server-verbose-logging" argument - but it's really only something you need whilst debugging.  Now the UI is there, that has the information anyway.

@corrados - any view?

**Comment by hartym** [comment]
I know that more central servers can be added, but while some of us are happy to tune those kind of things, a lot of our fellows musician friends just want to play together so I can't really explain to them to switch thinngs like central servers or so ... I wondered if the limit was something imposed by resource constraints or just something easily upgradeable on the central server side. Just suggesting.

Although I'm aware that's something only required while setting up or debugging things, it's an usual practice I think for a command acting as a client (yes, it's a server, but acting as a client towards the central server here, I guess) to say (at least in a verbose mode) "Attempting connection to ..." and "Connection successful", or "Connection failed: blah". I'd gladly add it if I can (my cpp skills were not used for like 15 years, but probably adding a log is not too much work, I'll try it and appreciate if you point me to files to look for), may save a lot of time to people not trying the gui server (I was not even aware the server could run with the gui, although --no-gui should have been an hint). 

@corrados Thank you for this great work, really saves us in this hard times :)

Cheers.

**Comment by pljones** [comment]
Yeah, I know what you mean.

There could be a "cross-registration" system of central servers (i.e. if you're a central server, you can publish yourself to others and then a client gets the full list to pull servers down from).

But it does mean the client then polls a growing list of servers, most of which would be of no interest as they're too far away in network terms.

So, the ideal is that you only use a central server publish servers that have a sensibly low ping to you - so the geographic split is good.  It's only just come in and it'll take time for server owners to upgrade and move to the NA server if it makes sense.

**Comment by corrados** [comment]
On the very first registration attempt we could always output the result on the command line if "no GUI server" is run. We do not need a command line argument for that. If a client connects, we already output that information on the command line, e.g. https://github.com/corrados/jamulus/blob/master/src/serverlogging.cpp#L66

@pljones can you please take care of it? It should be possible with just a few lines of code. But make sure it only outputs at the very first registration attempt and not the successive registrations every 15 minutes.

**Comment by corrados** [comment]
> I know that more central servers can be added, but while some of us are happy to tune those kind of things, a lot of our fellows musician friends just want to play together so I can't really explain to them to switch thinngs like central servers or so

If you live in the US or Canada, the Jamulus client is automatically configured to use the North America Central Server list. So if your friends start Jamulus for the first time, they need nothing to do to see the correct list.

**Comment by hartym** [comment]
So after bannging my head against C++, I managed to add a console log that happens whenever the ServerListManager's status changes, and it looks that the reason of the failed registration differs from the onne I got in gui : 

```ubuntu@scw-clever-blackburn:~/jamulus$ ./Jamulus -s -n -p 1666 -e jamulus.fischvolk.de -o "DO NOT USE - DEV SERVER;Paris;74"
- server mode chosen
- no GUI mode chosen
- selected port number: 1666
- central server: jamulus.fischvolk.de
- server info: DO NOT USE - DEV SERVER;Paris;74
Server RegStatus Changed: Registration requested
 *** Jamulus, Version 3.5.1
 *** Internet Jam Session Software
 *** Under the GNU General Public License (GPL)
Server RegStatus Changed: Registration requested
Server RegStatus Changed: Registration requested
Server RegStatus Changed: Registration requested
Server RegStatus Changed: Registration requested
Server RegStatus Changed: Registration failed```

For info, if the reason was server full, I should see it, as you can read in my commit : https://github.com/hartym/jamulus/commit/1cc399fd582db7189ce8096f76d451fb180c015f

For some reason, it appears that the related constant means "time out", not really sure why it does so.

**Comment by pljones** [comment]
That means
- the _name_ is known - it's been resolved
- the UDP request to register has been sent
- 500ms have been waited without a response
- another four times the above gets sent with no response
- so the software gives up

So it looks like some how any response sent back to you isn't routing correctly.  Given that I've had two slaves up connected from the same machine on the NA Central server, it's likely down to some issue in your data centre that only you can address, unfortunately.

The extra diagnostics to make it clearer (but need to be optional -- and the translation to string moved to a common place, probably `util.cpp`).

**Comment by corrados** [comment]
I close this issue as the problem mentioned in the title seems not to be an issue in the Jamulus software but seems to be an issue with the data centre.

**Comment by hartym** [comment]
@corrados 

I respect your decision to close the issue and won't reopen it myself, but I disagree with your conclusion.

Here are the reasons that make me think something goes unexpected in the jamulus software :

The exact same software version using the "gui" mode, with the exact same parameters (including the same port), on the exact same server, can register successfully with the same central server.

Also, if I do register the server using GUI, then kill it and launch it using the no-gui mode, then the server is still present in the central server list, and one can still connect using the "dangling" node announced previously, so the network is working perfectly here. (I register "TEST" using gui on port 1234, then kill and relaunch using nogui, and I can still connect to "TEST" using the jamulus client because IP/port are the same as before and I guess the garbage collection is done on a "slow" regular basis in the central server. I can then play music on this server although the no-gui server never could announce itself).

I will try to find some time to dig further in the issue adding traces in the code (or maybe with some tcpdump / strace), and if that's ok, will let you know what I find. Probably not before next week though.

**Comment by hartym** [comment]
Update : 
After stracing the network calls I believe that the reason of failure is the server description length : "123456789012345678901;Paris;74" never registers (systematic) while "12345678901234567890;Paris;74" (one less character) always register, or gets server full message.

So the issue title and original description is obviously wrong, but to me, that is indeed an issue with the software that should probably reject invalid input string ?

Having reverse enginered the problem, I think now that indeed, in protocol.cpp we have : 
```
    // server name
    if ( GetStringFromStream ( vecData,
                               iPos,
                               MAX_LEN_SERVER_NAME,
                               RecServerInfo.strName ) )
    {
        return true; // return error code
    }
```

and in protocol.h:
```
#define MAX_LEN_SERVER_NAME              20
```

Would a patch that reject the string (with a gentle error message) if its length is over `MAX_LEN_SERVER_NAME` in main.cpp something you'd accept ?

**Comment by corrados** [comment]
Ok, I just fixed that (see my recent commit).

**Comment by pljones** [comment]
> Edit : I somehow also manage to miss something in my diagnostic in previous message, I guess the server name provided in gui override the server name passed on command line, hence the difference although parameters stayed the same.

It took me a while to understand that the GUI overrides _everything_ pretty much - apart from the local port, which can't be specified.

So I guess that should be fixed - (a) by adding a warning if running with GUI and passing parameters and (b) by allowing the local port to be specified.

**Comment by vocobox** [comment]
It is worth mentionning that using a non default port does not prevent from beeing registered by the central server. However the server name is not displayed in the server list. In this case, the server name is NetworkAdress:PortNumber, e.g. 45.281.125.x:42124

**Comment by corrados** [comment]
I cannot reproduce this issue. I just started my Windows Jamulus server with -s -p 42124 and registered it at the central server. When I then start a client in parallel, I can see that the server name is correctly listed.
How did you setup your server? Also using the GUI? Or just command line arguments?

**Comment by vocobox** [comment]
Only command line arguments. But I would not say it is a critical problem, mainly that it is good to know the current behaviour. The server is running on Ubuntu and the version I started is 3.5.0.

**Comment by vocobox** [comment]
Not related, but I noticed that tonight the server list is longer than usual to be displayed. If there is a too high load on central server, I may start another central server if you want. If I may help, simply send me an email to martin dot jamcloud at gmail dot com.

**Comment by corrados** [comment]
I do not think the problem is the server. If that would be the case, nobody could use the Central Server to jam but it just works perfectly, even in the evening when most of the people are online and jamming. I guess it is something in your local network which is causing this.
